src/commonTest/kotlin/org/intellij/markdown/JsonSpecTestBase.kt

```diff
@@ -0,0 +1,63 @@
+package org.intellij.markdown
+
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import org.intellij.markdown.flavours.MarkdownFlavourDescriptor
+import org.intellij.markdown.html.HtmlGenerator
+import org.intellij.markdown.parser.MarkdownParser
+import kotlin.test.assertEquals
+
+@Serializable
+private data class SpecTest(
+        val markdown: String,
+        val html: String,
+        val example: Int,
+        val start_line: Int,
+        val end_line: Int,
+        val section: String,
+)
+
+abstract class JsonSpecTestBase : TestCase() {
+    abstract fun getTestDataPath(): String
+    abstract fun getFlavour(): MarkdownFlavourDescriptor
+
+    fun doTest(example: Int) {
+        val tests = Json { ignoreUnknownKeys = true }.decodeFromString<List<SpecTest>>(readFromFile(getTestDataPath()))
+        val test = tests.first { it.example == example }
+        val flavour = getFlavour()
+        val tree = MarkdownParser(flavour).buildMarkdownTreeFromString(test.markdown)
+        val html = HtmlGenerator(test.markdown, tree, flavour, false).generateHtml()
+        assertEquals(
+                normalizeHtml(test.html),
+                normalizeHtml(html).removeSurrounding("<body>", "</body>")
+        )
+    }
+}
+
+private val HTML_TAG_REGEX = Regex("<[^>]+>")
+
+/** Remove newlines from [html], unless they occur in a tag or within text enclosed by `<pre>` tags. */
+private fun normalizeHtml(html: String): String = buildString {
```

Comment:
The spec tests tend to put tags on separate lines, presumably to make the examples easier to read. The default html generator outputs minified html, so we need to do something like to to avoid spurious failures. Let me know if there's a better way to get the output to line up with the tests.